### PR TITLE
data_source/aws_ecr_lifecycle_policy_document: adding new data source for ECR

### LINKED ISSUE #6212: More than 2 ECR lifecycle policies can't be defined for one ECR repository
### Terraform Version

v0.11.7

### Affected Resource(s)

* aws_ecr_lifecycle_policy

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
resource "aws_ecr_repository" "sample" {
  name = "${format("%s-api",lookup(var.default, "name"))}"
}

resource "aws_ecr_lifecycle_policy" "sample_policy_1" {
  repository = "${aws_ecr_repository.sample.name}"

  policy = <<EOF
{
    "rules": [
        {
            "rulePriority": 1,
            "description": "sample policy 1",
            "selection": {
                "tagStatus": "tagged",
                "tagPrefixList": ["sample_tag_1"],
                "countType": "imageCountMoreThan",
                "countNumber": 10
            },
            "action": {
                "type": "expire"
            }
        }
    ]
}
EOF
}

resource "aws_ecr_lifecycle_policy" "sample_policy_2" {
  repository = "${aws_ecr_repository.sample.name}"

  policy = <<EOF
{
    "rules": [
        {
            "rulePriority": 2,
            "description": "sample policy 2",
            "selection": {
                "tagStatus": "tagged",
                "tagPrefixList": ["sample_tag_2"],
                "countType": "imageCountMoreThan",
                "countNumber": 10
            },
            "action": {
                "type": "expire"
            }
        }
    ]
}
EOF
}
```

#### For multiple ECR lifecycle policies in the same single .tfstafe file

All of the ECR lifecycle policies which were defined in a single .tf file shoud be attached to the specified ECR repository.

#### For multiple ECR lifecycle policies in multiple .tfstafe files

When I prepared two .tfstate files, which include ECR lifecycle policiy definitions for a single ECR repository, All of the ECR lifecycle policies are attached to the specified ECR repository, and all .tfstate files should keep them.

#### For multiple ECR lifecycle policies in the same single .tfstafe file

Just one ECR lifecycle policy that was applied at last was attached to the ECR repository. All of the other ECR lifecycle policies were once created, and overwritten by following ECR lifecycle policy deifnitions.

e.g.)
1. I defined ECR lifecycle policy `A` and `B` for `ECR repository a`.
2. When I applied the .tf file, policy `A` was attached to the `repository a`.
3. After that, policy `A` was overwritten by `B`, thus policy `A` was destroyed.
4. In the end, just only policy `B` was attached to the `repository a`.

#### For multiple ECR lifecycle policies in multiple .tfstafe files

e.g.)
**ECR lifecycle policy A** is defined in `sample_a.tf` and `sample_a.tfstate` for **ECR repository a**.
Also, **ECR lifecycle policy B** is defined in `sample_b.tf` and `sample_b.tfstafe` for **ECR repository a**, too.
When I applied `sample_a.tf`, **ECR lifecycle policy A** was correctly attached to **ECR repository a** . After that, I applied `sample_b.tf` and found that **ECR lifecycle policy A** was unexpectedly deleted and *ECR lifecycle policy B* was attached to **ECR repository a** .

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

Nothing

### References

* https://www.terraform.io/docs/providers/aws/r/ecr_lifecycle_policy.html

### PR DESCRIPTION
**Pull request by slapula**
Closes #6212 

In the spirit of `aws_iam_policy_document`, I tried my hand at doing the same thing for ECR lifecycle policy documents. 

Changes proposed in this pull request:

* Adds `aws_ecr_lifecycle_policy_document` data source

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSDataSourceEcrLifecyclePolicyDocument_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSDataSourceEcrLifecyclePolicyDocument_basic -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSDataSourceEcrLifecyclePolicyDocument_basic
--- PASS: TestAccAWSDataSourceEcrLifecyclePolicyDocument_basic (10.88s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	10.904s
```

### CODE REVIEW

**Comment by bflad** [comment]
Adding reference to #6212

**Comment by bflad** [comment]
I don't have time right now to give a full review of this PR, but it definitely seems like something we will be interested in implementing. 👍

File: aws/data_source_aws_ecr_lifecycle_policy_document.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"priority": {
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Required: false,
```
Comment (bflad) [inline aws/data_source_aws_ecr_lifecycle_policy_document.go]: Nit: `Required: false` is extraneous here and below.

File: aws/data_source_aws_ecr_lifecycle_policy_document.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+									"count_number": {
+										Type:     schema.TypeInt,
+										Required: false,
+										Optional: true,
+									},
+								},
+							},
+						},
+						"action": {
+							Type:     schema.TypeSet,
```
Comment (bflad) [inline aws/data_source_aws_ecr_lifecycle_policy_document.go]: This attribute should be `Type: schema.TypeList` and `MaxItems: 1` 👍

File: aws/data_source_aws_ecr_lifecycle_policy_document.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+		if description, ok := cfgStmt["description"]; ok {
+			stmt.Description = description.(string)
+		}
+
+		if selection := cfgStmt["selection"].(*schema.Set).List(); len(selection) > 0 {
+			stmt.Selection = dataSourceAwsEcrLifecyclePolicyDocumentMakeSelection(selection)
+		}
+
+		stmt.Action = EcrLifecyclePolicyAction{
```
Comment (bflad) [inline aws/data_source_aws_ecr_lifecycle_policy_document.go]: This should either attempt to read `cfgStmt["action"]` (and default to adding this configuration if its length is 0) or the `action` attribute should be removed.

**Comment by bflad** [comment]
Same here -- just please let us know if you don't have time to work on addressing the feedback.

File: aws/data_source_aws_ecr_lifecycle_policy_document.go (hunk 2 above)
Comment (slapula) [reply aws/data_source_aws_ecr_lifecycle_policy_document.go]: Fixed

File: aws/data_source_aws_ecr_lifecycle_policy_document.go (hunk 1 above)
Comment (slapula) [reply aws/data_source_aws_ecr_lifecycle_policy_document.go]: Both have been removed

File: website/docs/d/ecr_lifecycle_policy_document.html.markdown [hunk 4]
```diff
@@ -0,0 +1,56 @@
+---
+layout: "aws"
+page_title: "AWS: aws_ecr_lifecycle_policy_document"
+sidebar_current: "docs-aws-datasource-ecr-lifecycle-policy-document"
+description: |-
+    PGenerates an ECR lifecycle policy document in JSON format.
```
[the remark this answers is not in the document]
Comment (slapula) [reply website/docs/d/ecr_lifecycle_policy_document.html.markdown]: Corrected

File: website/docs/d/ecr_lifecycle_policy_document.html.markdown [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+---
+
+# Data Source: aws_ecr_lifecycle_policy_document
+
+This is a data source which can be used to construct a JSON representation of an ECR lifecycle policy document, for use with resources which expect policy documents, such as the aws_ecr_lifecycle_policy resource.
```
[the remark this answers is not in the document]
Comment (slapula) [reply website/docs/d/ecr_lifecycle_policy_document.html.markdown]: Updated

File: aws/data_source_aws_ecr_lifecycle_policy_document_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+		return nil
+	}
+}
+
+var testAccAWSEcrLifecyclePolicyDocumentConfig = `
+data "aws_ecr_lifecycle_policy_document" "test" {
+    rule {
+      priority        = 1
+			description     = "This is a test."
```
[the remark this answers is not in the document]
Comment (slapula) [reply aws/data_source_aws_ecr_lifecycle_policy_document_test.go]: Oops, not sure how this got that way.  I blame VSCode. :+1:

File: website/docs/d/ecr_lifecycle_policy_document.html.markdown [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+-> For more information about building AWS ECR lifecycle policy documents with Terraform, see the [AWS ECR Lifecycle Policy Document Guide](https://docs.aws.amazon.com/AmazonECR/latest/userguide/LifecyclePolicies.html).
```
[the remark this answers is not in the document]
Comment (slapula) [reply website/docs/d/ecr_lifecycle_policy_document.html.markdown]: Corrected

File: aws/data_source_aws_ecr_lifecycle_policy_document_test.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	// acceptance test, but just instantiating the AWS provider requires
+	// some AWS API calls, and so this needs valid AWS credentials to work.
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSEcrLifecyclePolicyDocumentConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckEcrLifecyclePolicyValue("data.aws_ecr_lifecycle_policy_document.test", "json",
```
[the remark this answers is not in the document]
Comment (slapula) [reply aws/data_source_aws_ecr_lifecycle_policy_document_test.go]: Oh, cool, I have removed the unneeded custom function.

File: aws/data_source_aws_ecr_lifecycle_policy_document.go (hunk 3 above)
Comment (slapula) [reply aws/data_source_aws_ecr_lifecycle_policy_document.go]: I think I follow.  Is this (1a4f6cc) what you are looking for here?

Comment (bflad) [reply aws/data_source_aws_ecr_lifecycle_policy_document.go]: Yes, thank you!

File: aws/data_source_aws_ecr_lifecycle_policy_document_test.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAWSDataSourceEcrLifecyclePolicyDocument_basic(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (slapula) [reply aws/data_source_aws_ecr_lifecycle_policy_document_test.go]: I just added a couple more tests: one that omits an optional attribute (`descriptions`) and one that generates a two rule policy.

**Comment by jufemaiz** [comment]
This looks great! What's the likelihood of it being merged?

**Comment by sdouche** [comment]
Hi. This data source will be useful to remove a lot of lines in ECR definitions.

**Comment by jufemaiz** [comment]
Merge conflicts :sadpand:

**Comment by nomeelnoj** [comment]
Any update on this? This would also really help us.

**Comment by TechIsCool** [comment]
Would really be nice to not have to use JSON to handle these when implementing modules.

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by duncz** [comment]
I appreciate it's a small part of Terraform but I'd imagine most organisations will want to use this with ECR... and most organisations using AWS Terraform will likely need ECR!

Could this please be merged?

**Comment by c-ameron** [comment]
Hello, any update on this? Thanks :)

**Comment by Luis-3M** [comment]
This is powerful - managing it via a JSON is a mess and very error-prone.
When are we aiming to release it ?

**Review by HAHGX — APPROVED** [review]

**Comment by ewbankkit** [comment]
@slapula Thanks for the contribution :tada: :clap:.
This work has been rolled into https://github.com/hashicorp/terraform-provider-aws/pull/36493.

**Comment by slapula** [comment]
wow I forgot this was still open!  Thanks for bringing this across the finish line @ewbankkit

**Pull request closed without merging**